src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,422 @@
+     */
+    function createOffer(
+        address tokenContract,
+        uint256 tokenId,
+        uint256 quantity,
+        address currency,
+        uint256 pricePerToken,
+        uint256 expiry
+    ) external returns (bytes32 offerId) {
+        // Check approved currency for offer
+        uint256 total = quantity * pricePerToken;
+        if (!_hasApprovedCurrency(currency, total, msg.sender)) {
+            revert InvalidCurrencyApproval(currency, total, msg.sender);
+        }
+        // Check quantity
+        bool isERC1155 = _tokenIsERC1155(tokenContract);
+        if ((isERC1155 && quantity == 0) || (!isERC1155 && quantity != 1)) {
+            revert InvalidQuantity();
+        }
+
+        offerId = _createOrder(false, tokenContract, tokenId, quantity, currency, pricePerToken, expiry);
+
+        emit OfferCreated(offerId, tokenContract, tokenId, quantity, currency, pricePerToken, expiry);
+
+        return offerId;
+    }
+
+    /**
+     * Create an order, either listing or offer.
+     * @param isListing True if the order is a listing, false if it is an offer.
+     * @param tokenContract The address of the token contract.
+     * @param tokenId The ID of the token.
+     * @param quantity The quantity of tokens to list.
+     * @param currency The address of the currency to list the token for.
+     * @param pricePerToken The price per token. Note this includes royalties.
+     * @param expiry The timestamp at which the listing expires.
+     * @return orderId The ID of the order.
+     */
+    function _createOrder(
+        bool isListing,
+        address tokenContract,
+        uint256 tokenId,
+        uint256 quantity,
+        address currency,
+        uint256 pricePerToken,
+        uint256 expiry
+    ) internal returns (bytes32 orderId) {
+        if (pricePerToken == 0) {
+            revert InvalidPrice();
+        }
+        // solhint-disable-next-line not-rely-on-time
+        if (expiry <= block.timestamp) {
+            revert InvalidExpiry();
+        }
+
+        Order memory order = Order({
+            isListing: isListing,
+            creator: msg.sender,
+            tokenContract: tokenContract,
+            tokenId: tokenId,
+            quantity: quantity,
+            currency: currency,
+            pricePerToken: pricePerToken,
+            expiry: expiry
+        });
+        orderId = hashOrder(order);
+
+        if (orders[orderId].creator != address(0)) {
+            // Collision
+            revert InvalidOrderId(orderId);
+        }
+        orders[orderId] = order;
+
+        return orderId;
+    }
+
+    /**
+     * Purchases a token.
+     * @param listingId The ID of the listing.
+     * @param quantity The quantity of tokens to purchase.
+     * @param additionalFees The additional fees to pay.
+     * @param additionalFeeRecievers The addresses to send the additional fees to.
+     * @dev Royalties are taken from the listing cost.
+     */
+    function acceptListing(
```

Comment:
Silly question, but why use two different methods here? The order knows if its an offer or a listing.

Reply:
For constency with `createListing/Offer`, which are different to limit the amount of branching due to different validations. 
I'm 50/50 about combining these and using `Order` throughout. What do you think?

Reply:
I think having less code is better, and makes the code easier to follow. If we want to have different methods, they just be a single line wrapper of an internal method with a parameter.

Reply:
Removed listing/offer functions. Now it's xOrder and xOrderBatch method sigs